File: csharp/step1_read_print.cs
using System;

namespace Mal
{
    internal class MAL
    {
        static int Main(string[] args)
        {
            return Repl();
        }

        static Value READ(string arg)
        {
            reader.Set(arg);
            return reader.Read_form();
        }

        static Value EVAL(Value arg) => arg;

        static string PRINT(Value arg)
        {
            return Printer.Pr_str(arg, true);
        }

        static string Rep(string arg)
        {
            return PRINT(EVAL(READ(arg)));
        }

        internal static int Repl()
        {
            for (; ;)
            {
                try
                {
                    Console.Write("user> ");
                    var str = Console.ReadLine();
                    if (str == null)
                        break;
                    var val = Rep(str);
                    if (val != null)
                        Console.WriteLine(val);
                }
                catch (MalException ex)
                {
                    Console.WriteLine("ERROR: {0}", ex.Message);
                }
            }

            return 0;
        }

        private static readonly Reader reader = new Reader("");
    }
}

File: cxx/step1_read_print.cxx
#include <iostream>

#include "types.h"
#include "reader.h"
#include "printer.h"
#include "gc.h"

namespace MAL
{

Value * READ(const std::string & arg)
{
   reader rdr(arg);

   return rdr.read_form();
}

Value * EVAL(Value * arg, EnvPtr env)
{
   return arg;
}

std::string PRINT(Value * arg)
{
   return pr_str(arg, true);
}

std::string rep(const std::string & arg)
{
   EnvPtr env;
   return PRINT(EVAL(READ(arg), env));
}

void repl(const char * prompt)
{
   std::string input;

   while (readline(prompt, input))
   {
      try
      {
         std::cout << rep(input) << std::endl;
      }
      catch (MalException & ex)
      {
         std::cout << "Error: " << ex.get_msg() << std::endl;
      }
   }
}

}

int main(int argc, char * argv[])
{
   MAL::GC::initialize(1024*1024);

   MAL::repl("user> ");
   return 0;
}
